substrate/zombienet-sdk/tests/staking_next/tests.rs

```diff
@@ -0,0 +1,380 @@
+	let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+	let mut rx = rx;
+	let mut tx = Some(tx);
+
+	loop {
+		tokio::select! {
+			rc_block = rc_blocks_sub.next() => {
+				if let Some(block) = rc_block {
+					let block = block?;
+					for event in block.events().await?.iter() {
+						let event = event?;
+						if event.pallet_name() == "StakingNextAhClient" {
+							match event.variant_name() {
+								"ValidatorSetReceived" => {
+									let ev = parse_new_validator_set_count(event.field_values()?).expect("ValidatorSetReceived event should be parsed");
+									test_state.on_ah_client_event(ev);
+								},
+								_ => {
+									log::debug!(
+										"ON RC: block: {}; pallet: StakingNextAhClient;  Event: {:?}; Fields: {:?}",
+										block.number(),
+										event.variant_name(),
+										event.field_values()?
+									);
+								}
+							}
+						}
+					}
+				}
+			}
+			ah_block = ah_blocks_sub.next() => {
+				if let Some(block) = ah_block {
+					let block = block?;
+					for event in block.events().await?.iter() {
+						let event = event?;
+
+						if event.pallet_name() == "StakingNextRcClient" {
+							match event.variant_name() {
+								"SessionReportReceived" => {
+									let fields = event.field_values()?;
+									let ev = parse_session_report_received(fields).expect("SessionReportReceived event should be parsed");
+									test_state.on_rc_client_event(ev, &mut tx);
+								},
+								_ => {
+									log::debug!(
+										"ON AH: block: {} StakingNextRcClient Event: {:?}; Fields: {:?}",
+										block.number(),
+										event.variant_name(),
+										event.field_values()?
+									);
+								}
+							}
+
+						} else if event.pallet_name() == "Staking" {
+							match event.variant_name() {
+								"PagedElectionProceeded" => {
+									log::debug!(
+										"ON AH: block: {} Staking PagedElectionProceeded Event: {:?}; Fields: {:?}",
+										block.number(),
+										event.variant_name(),
+										event.field_values()?
+									);
+									let ev = parse_paged_election_proceeded(event.field_values()?).expect("PagedElectionProceeded event should be parsed");
+									test_state.on_rc_client_event(ev, &mut tx);
+								},
+								_ => {
+									log::debug!(
+										"ON AH: block: {} Staking Event: {:?}; Fields: {:?}",
+										block.number(),
+										event.variant_name(),
+										event.field_values()?
+									);
+								}
+							}
+						} else if event.pallet_name() == "StakingNextAhClient" {
+							assert!(false, "StakingNextAhClient event on AH client");
+						}
+						else {
+							log::debug!("block: {} Event for pallet: {:?}; variant: {:?}; fields: {:?}", block.number(), event.pallet_name(), event.variant_name(), event.field_values()?);
+						}
+					}
+				}
+			}
+			_ = &mut rx => {
+				log::info!("Test finished");
+				break
+			}
+		}
+	}
+
+	Ok(())
+}
+
+// TODO: There should be a better way to do this.
```

Comment:
my guess would again be subxt typegen here.

Reply:
Here I agree 100% - I'm reworking this right now.